File: alpaca/templates/ruby/lib/http_client.rb
require "{{call .Fnc.underscore .Pkg.name}}/http_client/auth_handler"
require "{{call .Fnc.underscore .Pkg.name}}/http_client/error_handler"
require "{{call .Fnc.underscore .Pkg.name}}/http_client/request_handler"
require "{{call .Fnc.underscore .Pkg.name}}/http_client/response_handler"

module {{.Pkg.name}}

  module HttpClient

    # Main HttpClient which is used by Api classes
    class HttpClient

      attr_accessor :options, :headers

      @headers = {}

      def initialize(auth = {}, options = {})
{{if .Api.authorization.oauth}}
        if auth.is_a? String
          auth = { access_token: auth }
        end
{{end}}
        @options = {
          base: "{{.Api.base}}",{{with .Api.version}}
          api_version: "{{.}}",{{end}}
          user_agent: "alpaca/0.1.0 (https://github.com/pksunkara/alpaca)"
        }

        @options.update options

        @headers = {}

        @headers[:user_agent] = @options[:user_agent]

        if @options.has_key? :headers
          @headers.update @options[:headers]
          @options.delete :headers
        end

        @client = Faraday.new url: @options[:base]
      end

      def get(path, params = {}, options = {})
        request path, nil, "get", options.merge({ query: params })
      end

      def post(path, body = {}, options = {})
        request path, body, "post", options
      end

      def patch(path, body = {}, options = {})
        request path, body, "patch", options
      end

      def delete(path, body = {}, options = {})
        request path, body, "delete", options
      end

      def put(path, body = {}, options = {})
        request path, body, "put", options
      end

      # Intermediate function which does three main things
      #
      # - Transforms the body of request into correct format
      # - Creates the requests with give parameters
      # - Returns response body after parsing it into correct format
      def request(path, body, method, options)
        headers = {}

        options = @options.merge options

        if options.has_key? :headers
          headers = options[:headers]
          options.delete :headers
        end

        headers = @headers.merge headers

        options.delete :body

        body, headers = set_body body, headers, options

        response = create_request method, path, body, headers, options

        body = get_body response
      end

      # Creating a request with the given arguments
      #
      # If api_version is set, appends it immediately after host
      def create_request(method, path, body, headers, options)
        version = options.has_key?(:api_version) ? "/#{options[:api_version]}" : ""
{{if .Api.response.suffix}}
        # Adds a suffix (ex: ".html", ".json") to url
        suffix = options.has_key?(:response_type) ? options[:response_type] : "{{or .Api.response.formats.default "html"}}"
        path = "#{path}.#{suffix}"
{{end}}
        path = "#{version}#{path}"

        instance_eval <<-RUBY, __FILE__, __LINE__ + 1
          @client.#{method} path, body, headers
        RUBY
      end

      # Get response body in correct format
      def get_body(response)
        {{.Pkg.name}}::HttpClient::ResponseHandler.get_body response
      end

      # Set request body in correct format
      def set_body(body, headers, options)
        {{.Pkg.name}}::HttpClient::RequestHandler.set_body body, headers, options
      end

    end

  end

end

File: templates/php/lib/HttpClient/HttpClient.php
<?php

namespace {{call .Fnc.camelize .Pkg.Name}}\HttpClient;

use Guzzle\Http\Client as GuzzleClient;
use Guzzle\Http\ClientInterface;
use Guzzle\Http\Message\RequestInterface;

use {{call .Fnc.camelize .Pkg.Name}}\HttpClient\AuthHandler;
use {{call .Fnc.camelize .Pkg.Name}}\HttpClient\ErrorHandler;
use {{call .Fnc.camelize .Pkg.Name}}\HttpClient\RequestHandler;
use {{call .Fnc.camelize .Pkg.Name}}\HttpClient\Response;
use {{call .Fnc.camelize .Pkg.Name}}\HttpClient\ResponseHandler;

/**
 * Main HttpClient which is used by Api classes
 */
class HttpClient
{
    protected $options = array(
        'base'    => '{{.Api.base}}',{{with .Api.version}}
        'api_version' => '{{.}}',{{end}}
        'user_agent' => 'alpaca/{{.Api.alpaca_version}} (https://github.com/pksunkara/alpaca)'
    );

    protected $headers = array();

    public function __construct($auth = array(), array $options = array())
    {
{{if .Api.authorization.oauth}}
        if (gettype($auth) == 'string') {
            $auth = array('access_token' => $auth);
        }
{{else}}{{if .Api.authorization.header}}
        if (gettype($auth) == 'string') {
            $auth = array('http_header' => $auth);
        }
{{end}}{{end}}
        $this->options = array_merge($this->options, $options);

        $this->headers = array(
            'user-agent' => $this->options['user_agent'],
        );

        if (isset($this->options['headers'])) {
            $this->headers = array_merge($this->headers, array_change_key_case($this->options['headers']));
            unset($this->options['headers']);
        }

        $client = new GuzzleClient($this->options['base'], $this->options);
        $this->client = $client;

        $listener = array(new ErrorHandler(), 'onRequestError');
        $this->client->getEventDispatcher()->addListener('request.error', $listener);

        if (!empty($auth)) {
            $listener = array(new AuthHandler($auth), 'onRequestBeforeSend');
            $this->client->getEventDispatcher()->addListener('request.before_send', $listener);
        }
    }

    public function get($path, array $params = array(), array $options = array())
    {
        return $this->request($path, null, 'GET', array_merge($options, array('query' => $params)));
    }

    public function post($path, $body, array $options = array())
    {
        return $this->request($path, $body, 'POST', $options);
    }

    public function patch($path, $body, array $options = array())
    {
        return $this->request($path, $body, 'PATCH', $options);
    }

    public function delete($path, $body, array $options = array())
    {
        return $this->request($path, $body, 'DELETE', $options);
    }

    public function put($path, $body, array $options = array())
    {
        return $this->request($path, $body, 'PUT', $options);
    }

    /**
     * Intermediate function which does three main things
     *
     * - Transforms the body of request into correct format
     * - Creates the requests with give parameters
     * - Returns response body after parsing it into correct format
     */
    public function request($path, $body = null, $httpMethod = 'GET', array $options = array())
    {
        $headers = array();

        $options = array_merge($this->options, $options);

        if (isset($options['headers'])) {
            $headers = $options['headers'];
            unset($options['headers']);
        }

        $headers = array_merge($this->headers, array_change_key_case($headers));

        unset($options['body']);

        unset($options['base']);
        unset($options['user_agent']);
{{if .Api.no_verify_ssl}}
        $options['verify'] = false;
{{end}}
        $request = $this->createRequest($httpMethod, $path, null, $headers, $options);

        if ($httpMethod != 'GET') {
            $request = $this->setBody($request, $body, $options);
        }

        try {
            $response = $this->client->send($request);
        } catch (\LogicException $e) {
            throw new \ErrorException($e->getMessage());
        } catch (\RuntimeException $e) {
            throw new \RuntimeException($e->getMessage());
        }

        return new Response($this->getBody($response), $response->getStatusCode(), $response->getHeaders());
    }

    /**
     * Creating a request with the given arguments
     *
     * If api_version is set, appends it immediately after host
     */
    public function createRequest($httpMethod, $path, $body = null, array $headers = array(), array $options = array())
    {
        $version = (isset($options['api_version']) ? "/".$options['api_version'] : "");
{{if .Api.response.suffix}}
        // Adds a suffix (ex: ".html", ".json") to url
        $suffix = (isset($options['response_type']) ? $options['response_type'] : "{{or .Api.response.formats.default "html"}}");
        $path = $path.".".$suffix;
{{end}}
        $path    = $version.$path;

        return $this->client->createRequest($httpMethod, $path, $headers, $body, $options);
    }

    /**
     * Get response body in correct format
     */
    public function getBody($response)
    {
        return ResponseHandler::getBody($response);
    }

    /**
     * Set request body in correct format
     */
    public function setBody(RequestInterface $request, $body, $options)
    {
        return RequestHandler::setBody($request, $body, $options);
    }
}
